flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/SessionObserver.java

```diff
@@ -35,6 +37,19 @@ public SessionObserver(
 
     @Override
     public void observe(FlinkDeployment flinkApp, Context context, Configuration effectiveConfig) {
```

Comment:
Sorry for the late response. It seems that the current implementation could be optimized by a similar logic with `JobObserver#observe()`. Because the `JobManagerDeploymentStatus` could be set to `READY` in `observeJmDeployment`.

```
if (!isClusterReady(flinkApp)) {
      observeJmDeployment(flinkApp, context, effectiveConfig);
}

if (isClusterReady(flinkApp)) {
        // Check if session cluster can serve rest calls following our practice in JobObserver
        try {
            flinkService.listJobs(effectiveConfig);
        } catch (Exception e) {
            logger.error("REST service in session cluster is bad now", e);
            if (e instanceof TimeoutException) {
                // check for problems with the underlying deployment
                observeJmDeployment(flinkApp, context, effectiveConfig);
            }
        }
}
```